[#2487] exclude spam resources from discover

devincowan opened the pull request:
Resolves #2487 
<!--

Please read, and add your text at the bottom of this message.

Thanks for contributing code to HydroShare. In order to maintain code quality and expedite this process, please assist the development team by making sure the following is present in this pull request.

For more information, see https://docs.google.com/document/d/1dzxqlZW5fKNEyQSeKiSFq-SmS-VOPCva95XXkBjPExs

-->

### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

<!-- Enter steps that a QA engineer, stakeholder, or user documentation writer would follow to test the positive or "successful" case of the functionality your code provides or fixes -->

- [ ] Automated Testing

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Code coverage must not decrease so new functionality or code paths added during a bug fix must have appropriate tests written. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Sufficient User and Developer Documentation

<!-- Please email your positive test case lbrazil@cuahsi.org, who will make the decision regarding user documentation. -->

- [ ] Passing Jenkins Build

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Peer Code review and approval

<!-- This is the process by which a peer developer on the HydroShare team will read the changeset, provide feedback, and ultimately give a formal approval to the code before it passes PR status. -->

### Positive Test Case
1. Create a resource that contains one of the [spam patterns](https://github.com/CUAHSI/spam_patterns/blob/main/spam_patterns/patterns_re.py)
2. See that it does not show in Discover
3. Log in with an admin account
4. Click "allow in discover..."
<img width="350" alt="image" src="https://github.com/hydroshare/hydroshare/assets/17934193/0d5cab68-36fd-4dcc-8eb2-a6feddd01c0b">

5. Verify that the spam-filled resource has been allowlisted and shows up in Discover

devincowan commented on hs_core/models.py:
```diff
@@ -4108,6 +4136,11 @@ def show_in_discover(self):
             return False
+
+        if not self.spam_allowlisted:
+            if not self.free_of_spam_patterns:
```
the `show_in_discover` property is used during SOLR update in deployment. In my testing, the regex search takes .01 sec per resource so I'm hoping this doesn't slow down our deployments substantially

sblack-usu commented:
What is the plan for monitoring of resources marked as spam?  Are there plans to notify the user?

devincowan commented:
Wise question @sblack-usu! I think this will be a 2 pronged approach. 
1. We are spinning Neal's spam user detector app back up. App used to identify and remove spam users/resources. Manual process that takes some time before a spam resource is removed.
2. This PR aims to fill the gap -- keep spam resources from being discoverable in the interim before admin has had a chance to review within the detector app.

Re do we want to notify (in the UI) the owner of a dissalow-listed resource:
I vote "No" but let's talk.
Argument for: notification is to notify legit users that their resource isn't visible in Discover even though they tried to make it Discoverable.
Argument against: we are providing feedback to our spammers which will aid them in improving their spamming -> result in "arms race" of spam patterns -> increasingly complex regexs -> implementing AI spam detectors...

devincowan commented:
@sblack-usu this PR is ready for re-review, thank you!
Since your first look, I added:
* Management command for viewing resources containing spam patterns
* Made the spam patterns less aggressive (smaller net) to reduce false positives
* Added tests

sblack-usu commented on hs_core/management/commands/spam_resources.py:
```diff
@@ -0,0 +1,97 @@
+                    print("== THIS RESOURCE HAS BEEN ALLOWLISTED AND WILL SHOW UP IN DISCOVER ==")
+                else:
+                    shadow_banned_resources.append(res_info)
+                print(f"Checked {current_resource}/{total_res_to_check}")
```
Maybe print out the resource id too? You list the spam pattern, it may be nice to have the resource id that has the spam pattern.

devincowan replied:
good call, added that

sblack-usu reviewed (dismissed):
I made a couple of minor comments, overall this looks great.

devincowan commented on hs_discover/views.py:
```diff
@@ -115,6 +115,10 @@ def get(self, request, *args, **kwargs):
             if filters.get('geofilter'):
                 sqs = sqs.filter(north__range=[-90, 90])  # return resources with geographic data
 
+            # Possible to filter spam patterns here
```
thanks removed it

devincowan commented:
thanks for taking a look @sblack-usu! I made the changes you suggested
